notebooks/network-analysis.ipynb

```diff
@@ -0,0 +1,434 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Visualize the supernetwork\n",
+    "- To visualize the supernetwork, we need to convert the `dat_geo` DataFrame to a GeoDataFrame, named `gdf`.\n",
+    "- The GeoDataFrame contains geometry information that can be easily plotted.\n",
+    "- We make a simple map that represents all segments in the supernetwork as small points on a CONUS basemap. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# create a GeoDataFrame for viz purposes\n",
+    "gdf = geopandas.GeoDataFrame(dat_geo, geometry=geopandas.points_from_xy(dat_geo.lon, dat_geo.lat))\n",
+    "\n",
+    "# remote the parent DataFrame from memory\n",
+    "del dat_geo\n",
+    "\n",
+    "# import world basemap\n",
+    "world = geopandas.read_file(geopandas.datasets.get_path('naturalearth_lowres'))\n",
+    "\n",
+    "# create a basemap showing the boundaries of the USA\n",
+    "base = world[world.name == \"United States of America\"].plot(\n",
+    "    color='white', edgecolor='black')\n",
+    "\n",
+    "# add our stream network, segments are represented as small points (nodes in the network)\n",
+    "gdf.plot(ax = base, markersize = 0.01, color = \"black\")\n",
+    "\n",
+    "# crop the map extent\n",
+    "base.set_xlim(-130,-60)\n",
+    "base.set_ylim(20,55)\n",
+    "\n",
+    "# add a title and display\n",
+    "plt.title('The CONUS (mainstem) supernetwork')\n",
+    "plt.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Handle terminal segments\n",
+    "- In the native data, terminal segments (those that do not flow into another segment) have a downstream segment ID of zero in the downstream column of the DataFrame.\n",
+    "- The function `repace_downstreams` changes zeros to the negative of the terminal segment ID. For example, segment ID 12345 is a terminal segment, the downstream segment ID will be set to -12345."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# replace terminal donwstream nodes\n",
+    "def replace_downstreams(data, downstream_col, terminal_code):\n",
+    "    ds0_mask = data[downstream_col] == terminal_code # True if `downstream_col` == 0\n",
+    "    new_data = data.copy()\n",
+    "    new_data.loc[ds0_mask, downstream_col] = -ds0_mask.index[ds0_mask] #??\n",
+    "    return new_data\n",
+    "\n",
+    "# call function above...\n",
+    "data = replace_downstreams(data, network_data['downstream_col'], 0)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Extract network connections\n",
+    "- Upstream and downstream network connections are extracted with the `extract_network` and `reverse_network` functions\n",
```

Comment:
This should read more clearly, "Downstream and upstream network connections ..."